Improve UI for creating and editing plugins.

EricEzaM opened the pull request:
Commit 1: Moves `set_as_minsize()` from `Popup` to `Window`. This method only invokes methods in `Window`, and it is a nice helper to have on `Window` instead of just `Popup`, ~and I have used it in Commit 2.~

Commit 2: Improves plugin creation dialog by reporting validation errors in the form. Closes #41736

![plugin_creation_dialog_final](https://user-images.githubusercontent.com/41730826/92338271-9d4c5a00-f0f2-11ea-9d75-6e6fc4b75a09.gif)

EricEzaM commented:
Oh I also changed `tool` to `@tool` to match GDScript 2.0 convention. I know this section of the file is a `TODO` for a rework but this is just so at least it doesn't throw error when you made a plugin.

jkb0o commented:
IMO, changing the size of the popup because of the number of warnings isn't leading to good UX. It would be cool if the box with warnings keeps constant minimum size and just does not grow. I guess, as far as all warnings should be fixed to create the plugin, it's ok if we omit messages that don't fit.

EricEzaM commented:
Not sure if this is really cherry pickable for 3.2 since in 3.2 Popup does not inherit from Window.

@jkb0o I made it so because to me it seems strange to have a large void of nothing when there are no errors.

jkb0o commented:
When large void of nothing fights unnatural twitching, I bet on the first one. But it is not the only opinion. Anyway, it is nice to see warnings reported this way (inside dialog).

Calinou commented:
I agree it would be better to reserve some space for errors too. We can use the same pattern as the script creation dialog where a Panel is used to contain error messages (with a darker background).

MarianoGnu commented:
@EricEzaM instead of showing errors on the bottom you could show error/warning icons next to the field with problems, and click or hover it could display the actual error, something like this;

![image](https://user-images.githubusercontent.com/1776044/92306815-6ba78780-ef68-11ea-91d1-5b18cc6cb252.png)

EricEzaM commented:
Here is what I have at the moment, however I am considering (and have requested feedback regarding) abstracting this feature and making something like a `ValidatedLineEdit` control which is only used in the editor.

![validation](https://user-images.githubusercontent.com/41730826/92317512-bb5b8100-f044-11ea-9434-a30f49f47d09.gif)

This could be useful to replace similar instances of this sort of UI, as below, and also make form-like user interface elements easier to implement in the future.
![image](https://user-images.githubusercontent.com/41730826/92317521-d3cb9b80-f044-11ea-8071-a0eb7c9552f1.png)
![image](https://user-images.githubusercontent.com/41730826/92317524-dded9a00-f044-11ea-94b0-ca4255f56069.png)

EricEzaM commented:
After further experimentation I don't think it is the best option to make some general `Validated` control. At least not on a small scale. So for now I think this should be implemented on an ad-hoc basis. Below is the final product. If this is merged I will also port this functionality to the script creation dialog and the project settings advanced menu, if desired.

![proj_settings_done](https://user-images.githubusercontent.com/41730826/92321488-c4147d00-f06d-11ea-8a0b-613a831a1cf2.gif)

Calinou commented:
@EricEzaM There's the `StatusSuccess` icon you can use instead of the CheckBox node icon.

EricEzaM commented:
@Calinou don't know how I missed that StatusSuccess icon lol

![plugin_creation_dialog_final](https://user-images.githubusercontent.com/41730826/92338271-9d4c5a00-f0f2-11ea-9d75-6e6fc4b75a09.gif)

Jummit commented:
Note that some of this functionality would be obsolete if #36677 is merged.

Calinou commented:
I thought #36677 was merged already :slightly_frowning_face:

EricEzaM commented:
@Jummit It wouldn't be completely obsolete. Just because the fields are optional does not mean that they do not require validation if they _are_ filled in. If the user decides to set those fields then the values still need to be valid.

Additionally, if the plugin subfolder was still to match an existing folder, then the existing folder contents would still be overwriten which is undesirable.

If that pr was merged then the logic would just need to change so that blank script name = valid. If non blank, the validation would still need to occur. Everything else would stay as far as I can see.

Calinou reviewed (commented):
Should be good to merge after rebasing. I'll salvage changes from #36677 that are still relevant once this PR is merged.

**Note to maintainers:** If there are plans to cherry-pick this to the `3.x` branch, make sure not to cherry-pick the `@tool` change :slightly_smiling_face:

EricEzaM committed:
Added validation to plugin creation.

Improved plugin creation dialog by adding form error reporting and restricting the user from overwriting an existing folder.

akien-mga commented:
Thanks!
